xen/drivers/passthrough/arm/ipmmu-vmsa-plat.c

```diff
@@ -0,0 +1,259 @@
+
+/* SYSC Status Register */
+#define SYSCSR_PONENB		1	/* Ready for power resume requests */
+
+/* Power Control Register Offsets inside the register block for each domain */
+#define PWRSR_OFFS		0x00	/* Power Status Register */
+#define PWRONCR_OFFS	0x0c	/* Power Resume Control Register */
+#define PWRER_OFFS		0x14	/* Power Shutoff/Resume Error */
+
+#define SYSCSR_RETRIES		1000
+#define SYSCSR_DELAY_US		10
+
+#define PWRER_RETRIES		1000
+#define PWRER_DELAY_US		10
+
+#define SYSCISR_RETRIES		1000
+#define SYSCISR_DELAY_US	10
+
+struct rcar_sysc_ch {
+	const char *name;
+	u16 chan_offs;		/* Offset of PWRSR register for this area */
+	u8 chan_bit;		/* Bit in PWR* (except for PWRUP in PWRSR) */
+	u8 isr_bit;			/* Bit in SYSCI*R */
+};
+
+/*
+ * For the most of IPMMU-XX which are located in ALWAYS_ON power domain
+ * we don't care at all. But some of them are located in other domains
+ * and must be turned on once at boot.
+ * Hopefully, the each of domains we are dealing with within this file
+ * (A3VP, A3VP) is identically configured across all SoCs (H3, M3 and M3N).
+ * This allow us not to introduce support for each SoC separately.
+ */
+static const struct rcar_sysc_ch rcar_sysc_chs[2] = {
+	{
+		.name = "A3VP",
+		.chan_offs = 0x340,
+		.chan_bit = 0,
+		.isr_bit = RCAR_GEN3_PD_A3VP,
+	},
+	{
+		.name = "A3VC",
+		.chan_offs = 0x380,
+		.chan_bit = 0,
+		.isr_bit = RCAR_GEN3_PD_A3VC,
+	},
+};
+
+#define dev_name(dev) dt_node_full_name(dev_to_dt(dev))
+
+static int __init rcar_sysc_init(void)
+{
+	u32 syscier, syscimr;
+	int i;
+
+	if (rcar_sysc_base)
+		return 0;
+
+	rcar_sysc_base = ioremap_nocache(RCAR_SYSC_BASE, RCAR_SYSC_SIZE);
+	if (!rcar_sysc_base) {
+		printk("failed to map SYSC MMIO range\n");
+		return -ENOMEM;
+	}
+
+	syscier = 0;
+	for (i = 0; i < ARRAY_SIZE(rcar_sysc_chs); i++)
+		syscier |= BIT(rcar_sysc_chs[i].isr_bit);
+
+	/*
+	 * Mask all interrupt sources to prevent the CPU from receiving them.
+	 * Make sure not to clear reserved bits that were set before.
+	 */
+	syscimr = readl(rcar_sysc_base + SYSCIMR);
+	syscimr |= syscier;
+	writel(syscimr, rcar_sysc_base + SYSCIMR);
+
+	/* SYSC needs all interrupt sources enabled to control power */
+	writel(syscier, rcar_sysc_base + SYSCIER);
+
+	return 0;
+}
+
+static bool __init rcar_sysc_power_is_off(const struct rcar_sysc_ch *sysc_ch)
+{
+	unsigned int status;
+
+	rcar_sysc_init();
+
+	status = readl(rcar_sysc_base + sysc_ch->chan_offs + PWRSR_OFFS);
+	if (status & BIT(sysc_ch->chan_bit))
+		return true;
+
+	return false;
+}
+
+static int __init rcar_sysc_power_on(const struct rcar_sysc_ch *sysc_ch)
+{
+	unsigned int status;
+	int ret = 0, i, j;
+
+	rcar_sysc_init();
+
+	writel(BIT(sysc_ch->isr_bit), rcar_sysc_base + SYSCISCR);
+
+	/* Submit power resume request until it was accepted */
+	for (i = 0; i < PWRER_RETRIES; i++) {
+
+		/* Wait until SYSC is ready to accept a power request */
+		for (j = 0; j < SYSCSR_RETRIES; j++) {
```

Comment:
I think such retry loops should be written as following:

```
retires_left = SYSCSR_RETRIES;
while (! (readl(rcar_sysc_base + SYSCSR) & BIT(SYSCSR_PONENB)) {
    udelay(SYSCSR_DELAY_US);
    if (retries_left-- <= 0)
         return -EAGAIN;
}

```
or

```
num_retries = 0;
while (! (readl(rcar_sysc_base + SYSCSR) & BIT(SYSCSR_PONENB)) {
    udelay(SYSCSR_DELAY_US);
    if (num_retries++ > SYSCSR_RETRIES )
         return -EAGAIN;
}
```

This is much shorter and this will ensure that you compare against `SYSCSR_RETRIES` exactly in one place.

Reply:
You know, I was trying the retain code as it was done in Linux for sync with the driver changes in the future easily.

Reply:
If you insist on changing I will change. But I would prefer to leave as is.

Reply:
I'm okay with this. Just showing how to do it  more error-proof.